Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt

```diff
@@ -1,42 +1,111 @@
+import kotlinx.coroutines.plus
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class RecoveryCertificateRepository @Inject constructor(
     @AppScope private val appScope: CoroutineScope,
-    private val dispatcherProvider: DispatcherProvider,
+    dispatcherProvider: DispatcherProvider,
     private val qrCodeExtractor: DccQrCodeExtractor,
-    valueSetsRepository: ValueSetsRepository,
+    private val storage: RecoveryCertificateStorage,
 ) {
 
-    val certificates: Flow<Set<RecoveryCertificateWrapper>> = flowOf(emptySet())
+    private val internalData: HotDataFlow<Set<RecoveryCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+    ) {
+        storage.recoveryCertificates
+            .map { recoveryCertificate ->
+                RecoveryCertificateContainer(
+                    data = recoveryCertificate,
+                    qrCodeExtractor = qrCodeExtractor
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored recovery certificate data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing data.") }
+            .onEach { recoveryCertificates ->
+                Timber.tag(TAG).v("Recovery Certificate data changed: %s", recoveryCertificates)
+                storage.recoveryCertificates = recoveryCertificates.map { it.data }.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot recovery certificate data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.IO)
+    }
 
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> =
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(InvalidRecoveryCertificateException::class)
     suspend fun registerCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("registerCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        val newContainer = qrCode.toContainer()
+        internalData.updateBlocking {
+            if (any { it.containerId == newContainer.containerId }) {
+                throw InvalidRecoveryCertificateException(
+                    InvalidHealthCertificateException.ErrorCode.ALREADY_REGISTERED
+                )
+            }
+            plus(newContainer)
+        }
+        return newContainer
     }
 
-    suspend fun deleteCertificate(containerId: RecoveryCertificateContainerId): RecoveryCertificateContainer? {
+    private fun RecoveryCertificateQRCode.toContainer() = RecoveryCertificateContainer(
+        data = StoredRecoveryCertificateData(
+            identifier = uniqueCertificateIdentifier,
```

Comment:
If the identifier is the `uniqueCertificateIdentifier` from the certificate, then we don't need to store it extra.